gctscript/vm/manager.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/thrasher-corp/gocryptotrader/engine/subsystem"
+	"github.com/thrasher-corp/gocryptotrader/log"
+)
+
+const gctscriptManagerName = "GCTScript"
+
+// GctScriptManager loads and runs GCT Tengo scripts
+type GctScriptManager struct {
+	config   *Config
+	started  int32
+	stopped  int32
+	shutdown chan struct{}
+	// Optional values to override stored config ('nil' if not overridden)
+	Enabled            *bool
+	MaxVirtualMachines *uint8
+}
+
+// Started returns if gctscript manager subsystem is started
+func (g *GctScriptManager) Started() bool {
+	return atomic.LoadInt32(&g.started) == 1
+}
+
+// Start starts gctscript subsystem and creates shutdown channel
+func (g *GctScriptManager) Start(config *Config, wg *sync.WaitGroup) (err error) {
+	g.config = config
+	if !g.IsEnabled() {
+		return errors.New("manager has been disabled")
+	}
+
+	if atomic.AddInt32(&g.started, 1) != 1 {
+		return fmt.Errorf("%s %s", gctscriptManagerName, subsystem.ErrSubSystemAlreadyStarted)
+	}
+
+	defer func() {
+		if err != nil {
+			atomic.CompareAndSwapInt32(&g.started, 1, 0)
+		}
+	}()
+	log.Debugln(log.Global, gctscriptManagerName, subsystem.MsgSubSystemStarting)
+
+	g.shutdown = make(chan struct{})
+	go g.run(wg)
+	return nil
+}
+
+// Stop stops gctscript subsystem along with all running Virtual Machines
+func (g *GctScriptManager) Stop() error {
+	if atomic.LoadInt32(&g.started) == 0 {
+		return fmt.Errorf("%s %s", gctscriptManagerName, subsystem.ErrSubSystemNotStarted)
+	}
+
+	if atomic.AddInt32(&g.stopped, 1) != 1 {
+		return fmt.Errorf("%s %s", gctscriptManagerName, subsystem.ErrSubSystemAlreadyStopped)
+	}
+
+	log.Debugln(log.GCTScriptMgr, gctscriptManagerName, subsystem.MsgSubSystemShuttingDown)
+	close(g.shutdown)
+	err := g.ShutdownAll()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (g *GctScriptManager) run(wg *sync.WaitGroup) {
+	log.Debugln(log.Global, gctscriptManagerName, subsystem.MsgSubSystemStarted)
+
+	wg.Add(1)
+	SetDefaultScriptOutput()
+	g.autoLoad()
+	defer func() {
+		atomic.CompareAndSwapInt32(&g.stopped, 1, 0)
+		atomic.CompareAndSwapInt32(&g.started, 1, 0)
+		wg.Done()
+		log.Debugln(log.GCTScriptMgr, gctscriptManagerName, subsystem.MsgSubSystemShutdown)
+	}()
+
+	<-g.shutdown
+}
+
+func (g *GctScriptManager) autoLoad() {
+	for x := range g.config.AutoLoad {
+		temp := g.New()
+		if temp == nil {
+			log.Errorf(log.GCTScriptMgr, "Unable to create Virtual Machine, autoload failed for: %v",
+				g.config.AutoLoad[x])
+			continue
+		}
+		var name = g.config.AutoLoad[x]
+		if filepath.Ext(name) != common.GctExt {
+			name += common.GctExt
+		}
+		scriptPath := filepath.Join(ScriptPath, name)
+		err := temp.Load(scriptPath)
+		if err != nil {
+			log.Errorf(log.GCTScriptMgr, "%v failed to load: %v", filepath.Base(scriptPath), err)
+			continue
+		}
+		go temp.CompileAndRun()
+	}
+}
+
+// IsEnabled returns whether the manager is enabled
+func (g *GctScriptManager) IsEnabled() bool {
```

Comment:
This function needs data race protection, we have GRPC CLI functions that can enable and disable the use of scripts and a data race can occur. Having this dedicated function to check is good, but the value can be changed while reading.

Reply:
I'm now thinking we don't actually need the extra enable flag and just use `Started()` to check this...
It's simpler if the `Enabled` flag is only considered on the Engine level to decide whether to start the script manager and other decisions are based on the run status, not the enablement.